File: src/main/java/me/partlysanestudios/partlysaneskies/data/skyblockdata/SkyblockDataManager.java
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//

package me.partlysanestudios.partlysaneskies.data.skyblockdata;

import com.google.gson.*;
import me.partlysanestudios.partlysaneskies.PartlySaneSkies;
import me.partlysanestudios.partlysaneskies.data.pssdata.PublicDataManager;
import me.partlysanestudios.partlysaneskies.system.requests.Request;
import me.partlysanestudios.partlysaneskies.system.requests.RequestsManager;

import java.io.IOException;
import java.net.MalformedURLException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class SkyblockDataManager {

//    --------------------------- Items ---------------------------
    public static ArrayList<String> bitIds = new ArrayList<>();
    private static HashMap<String, String> nameToIdMap = new HashMap<>();
    private static HashMap<String, SkyblockItem> idToItemMap = new HashMap<>();
    private static long lastAhUpdateTime = PartlySaneSkies.getTime();

    public static boolean checkLastUpdate() {
        return PartlySaneSkies.getTime() >= lastAhUpdateTime + (1000 * 60 * 5);
    }

    public static void updateBz() {
        try{
            RequestsManager.newRequest(new Request("https://api.hypixel.net/skyblock/bazaar", s -> {
                if (!s.hasSucceeded()) {
                    return;
                }
                JsonObject object = new JsonParser().parse(s.getResponse()).getAsJsonObject().getAsJsonObject("products");


                for (Map.Entry<String, SkyblockItem> en : idToItemMap.entrySet()) {
                    if (!object.has(en.getKey())) {
                        continue;
                    }
                    en.getValue().setBazaarSellPrice(object.getAsJsonObject(en.getKey()).getAsJsonObject("quick_status").get("sellPrice").getAsDouble());
                    en.getValue().setBazaarBuyPrice(object.getAsJsonObject(en.getKey()).getAsJsonObject("quick_status").get("buyPrice").getAsDouble());
                }
            }));


        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void updateAverageLowestBin() {
        try {
            RequestsManager.newRequest(new Request("https://moulberry.codes/auction_averages_lbin/1day.json", s -> {
                if (!s.hasSucceeded()) {
                    return;
                }
                @SuppressWarnings("unchecked")
                HashMap<String, Double> map = (HashMap<String, Double>) new Gson().fromJson(s.getResponse(),
                        HashMap.class);

                for (Map.Entry<String, SkyblockItem> en : idToItemMap.entrySet()) {
                    if (!map.containsKey(en.getKey())) {
                        continue;
                    }
                    en.getValue().setAverageLowestBinPrice(map.get(en.getKey()));
                }
            }));

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void updateAll() {
        lastAhUpdateTime = PartlySaneSkies.getTime();
        updateAverageLowestBin();
        updateBz();
        updateLowestBin();

    }

    public static void updateLowestBin() {
        try {
            RequestsManager.newRequest(new Request("http://moulberry.codes/lowestbin.json", s -> {
                if (!s.hasSucceeded()) {
                    return;
                }
                @SuppressWarnings("unchecked")
                HashMap<String, Double> map = (HashMap<String, Double>) new Gson().fromJson(s.getResponse(),
                        HashMap.class);

                for (Map.Entry<String, SkyblockItem> en : idToItemMap.entrySet()) {
                    if (!map.containsKey(en.getKey())) {
                        continue;
                    }
                    en.getValue().setLowestBinPrice(map.get(en.getKey()));
                }
            }));

        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    public static void initItems() throws IOException {
        RequestsManager.newRequest(new Request("https://api.hypixel.net/resources/skyblock/items", s -> {
            String itemDataString = s.getResponse();
            if (!s.hasSucceeded()) {

                return;
            }
            JsonObject itemDataJson = new JsonParser().parse(itemDataString).getAsJsonObject();

            JsonArray itemArray = itemDataJson.get("items").getAsJsonArray();

            nameToIdMap = new HashMap<>();
            idToItemMap = new HashMap<>();

            for (JsonElement itemJson : itemArray) {
                JsonObject itemObject = itemJson.getAsJsonObject();

                String id = itemObject.get("id").getAsString();
                String name = itemObject.get("name").getAsString();
                double npcSellPrice = -1;

                if (itemObject.has("npc_sell_price")) {
                    npcSellPrice = itemObject.get("npc_sell_price").getAsDouble();
                }

                String rarity = "COMMON";
                if (itemObject.has("tier")) {
                    rarity = itemObject.get("tier").getAsString();
                }


                SkyblockItem item = new SkyblockItem(id, name, npcSellPrice, rarity);
                nameToIdMap.put(name, id);
                idToItemMap.put(id, item);
            }
        }));

    }

    public static void initBitValues() throws IOException {
        JsonObject bitsShopObject = new JsonParser().parse(PublicDataManager.getFile("constants/bits_shop.json")).getAsJsonObject().getAsJsonObject("bits_shop");
        for (Map.Entry<String, JsonElement> entry : bitsShopObject.entrySet()) {
            String id = entry.getKey();
            int bitCost = entry.getValue().getAsInt();
            SkyblockItem item = getItem(id);
            if (item == null) {
                continue;
            }
            bitIds.add(item.getId());
            item.setBitCost(bitCost);
        }

    }

    public static String getId(String name) {
        return nameToIdMap.get(name);
    }

    public static SkyblockItem getItem(String id) {
        return idToItemMap.get(id);
    }

    public static void runUpdater() {
        if (checkLastUpdate()) {
            lastAhUpdateTime = PartlySaneSkies.getTime();
            new Thread(() -> {
                lastAhUpdateTime = PartlySaneSkies.getTime();
                updateAll();
                lastAhUpdateTime = PartlySaneSkies.getTime();
            }).start();
            lastAhUpdateTime = PartlySaneSkies.getTime();
        }

    }









    //    --------------------------- Skills ---------------------------
    private static HashMap<String, SkyblockSkill> idToSkillMap = new HashMap<>();
    public static void initSkills() throws MalformedURLException {
        RequestsManager.newRequest(new Request("https://api.hypixel.net/resources/skyblock/skills", s -> {
            String itemDataString = s.getResponse();
            if (!s.hasSucceeded()) {
                return;
            }
            JsonObject skillDataJson = new JsonParser().parse(itemDataString).getAsJsonObject();

            JsonObject skillObject = skillDataJson.get("skills").getAsJsonObject();

            idToSkillMap = new HashMap<>();

//            Goes through each skill
            for (Map.Entry<String, JsonElement> en : skillObject.entrySet()) {
//                Gets the id from the key of the element
                String id = en.getKey();
//                Gets the data from the skill
                JsonObject object = en.getValue().getAsJsonObject();

//                Gets the max level
                int maxLevel = object.get("maxLevel").getAsInt();
//                Gets the json array containing the level
                JsonArray levelArray = object.getAsJsonArray("levels");
//                Creates a new hashmap to store the level and the total Experience Required to get to that level
                HashMap<Integer, Float> levelToExpMap = new HashMap<>();
//                Adds the level to the map
                for (int i = 0; i < levelArray.size(); i++) {
                    int level = i + 1;
                    float experience = levelArray.get(i).getAsJsonObject().get("totalExpRequired").getAsFloat();

                    levelToExpMap.put(level, experience);
                }
//                Instantiates a new skyblock skill with all the data
                SkyblockSkill skill = new SkyblockSkill(id, maxLevel, levelToExpMap);
//                Adds the skill to the map
                idToSkillMap.put(id, skill);
            }
        }));
    }

    public static SkyblockSkill getSkill(String skillId) {
        return idToSkillMap.get(skillId);
    }


    //    --------------------------- Players ---------------------------
    private static final HashMap<String, SkyblockPlayer> playerCache = new HashMap<>();

    public static SkyblockPlayer getPlayer(String username) throws MalformedURLException {
        SkyblockPlayer player;
        if (playerCache.containsKey(username)) {
            player = playerCache.get(username);

            if (!player.isExpired()) {
                return player;
            }

        }
        else {
            player = new SkyblockPlayer(username);

        }
        player.instantiatePlayer();
        playerCache.put(username, player);
        return player;
    }
}

File: src/main/kotlin/me/partlysanestudios/partlysaneskies/data/skyblockdata/SkyblockDataManager.kt
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//

package me.partlysanestudios.partlysaneskies.data.skyblockdata

import com.google.gson.JsonParser
import me.partlysanestudios.partlysaneskies.PartlySaneSkies.Companion.config
import me.partlysanestudios.partlysaneskies.PartlySaneSkies.Companion.time
import me.partlysanestudios.partlysaneskies.data.api.Request
import me.partlysanestudios.partlysaneskies.data.api.RequestRunnable
import me.partlysanestudios.partlysaneskies.data.api.RequestsManager.newRequest
import me.partlysanestudios.partlysaneskies.data.pssdata.PublicDataManager.getFile
import me.partlysanestudios.partlysaneskies.events.SubscribePSSEvent
import me.partlysanestudios.partlysaneskies.events.data.LoadPublicDataEvent
import java.io.IOException
import java.net.MalformedURLException

object SkyblockDataManager {
    @SubscribePSSEvent
    fun onDataLoad(event: LoadPublicDataEvent?) {
        try {
            initBitValues()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    //    --------------------------- Items ---------------------------
    var bitIds = ArrayList<String>()
    private var nameToIdMap = HashMap<String, String>()
    private var idToItemMap = HashMap<String, SkyblockItem>()
    private var lastAhUpdateTime = time
    private fun checkLastUpdate(): Boolean {
        return time >= lastAhUpdateTime + 1000 * 60 * 5
    }

    fun updateAll() {
        lastAhUpdateTime = time
        updateItems()
    }

    @Throws(IOException::class)
    private fun updateItems() {
        newRequest(
            Request("${config.apiUrl}/v1/hypixel/skyblockitem",
                RequestRunnable { s: Request ->
                    val itemDataString = s.getResponse()
                    if (!s.hasSucceeded()) {
                        return@RequestRunnable
                    }
                    val itemDataJson = JsonParser().parse(itemDataString).getAsJsonObject()
                    val products = itemDataJson.getAsJsonArray("products")

                    for (product in products) {
                        val en = product.asJsonObject
                        val rarity = try {
                            Rarity.valueOf(en.get("rarity")?.asString ?: "")
                        } catch (e: IllegalArgumentException) {
                            Rarity.UNKNOWN
                        }

                        val skyblockItem = SkyblockItem(
                            en.get("itemId").asString,
                            rarity,
                            en.get("name")?.asString ?: "",
                            en.get("npcSell")?.asDouble ?: 0.0,
                            en.get("bazaarBuy")?.asDouble ?: 0.0,
                            en.get("bazaarSell")?.asDouble ?: 0.0,
                            en.get("averageBazaarBuy")?.asDouble ?: 0.0,
                            en.get("averageBazaarSell")?.asDouble ?: 0.0,
                            en.get("lowestBin")?.asDouble ?: 0.0,
                            en.get("averageLowestBin")?.asDouble ?: 0.0,
                            en.get("material")?.asString ?: "",
                            en.get("unstackable")?.asBoolean ?: false
                        )

                        idToItemMap[en.get("itemId").asString] = skyblockItem
                        nameToIdMap[en.get("name").asString] = en.get("itemId").asString

                    }

                }, inMainThread = false, executeOnNextFrame = false, acceptAllCertificates = false
            )
        )
    }

    fun initBitValues() {
        val bitsShopObject =
            JsonParser().parse(getFile("constants/bits_shop.json")).getAsJsonObject().getAsJsonObject("bits_shop")
        for ((id, value) in bitsShopObject.entrySet()) {
            val bitCost = value.asInt
            val item = getItem(id)
                ?: continue
            bitIds.add(item.id)
            item.bitCost = bitCost
        }
    }

    fun getId(name: String): String {
        return if (!nameToIdMap.containsKey(name)) {
            ""
        } else nameToIdMap[name]!!
    }

    fun getItem(id: String): SkyblockItem? {
        return if (!idToItemMap.containsKey(id)) {
            null
        } else idToItemMap[id]
    }

    fun runUpdaterTick() {
        if (checkLastUpdate()) {
            lastAhUpdateTime = time
            Thread {
                lastAhUpdateTime =
                    time
                updateAll()
                lastAhUpdateTime =
                    time
            }.start()
            lastAhUpdateTime = time
        }
    }

    //    --------------------------- Skills ---------------------------
    private var idToSkillMap = HashMap<String, SkyblockSkill>()

    @Throws(MalformedURLException::class)
    fun initSkills() {
        newRequest(
            Request(
                "https://api.hypixel.net/resources/skyblock/skills",
                RequestRunnable { s: Request ->
                    val itemDataString = s.getResponse()
                    if (!s.hasSucceeded()) {
                        return@RequestRunnable
                    }
                    val skillDataJson =
                        JsonParser().parse(itemDataString).getAsJsonObject()
                    val skillObject = skillDataJson["skills"].getAsJsonObject()
                    idToSkillMap =
                        HashMap()

//            Goes through each skill
                    for ((id, value) in skillObject.entrySet()) {
//                Gets the id from the key of the element
                        //                Gets the data from the skill
                        val obj = value.getAsJsonObject()

//                Gets the max level
                        val maxLevel = obj["maxLevel"].asInt
                        //                Gets the json array containing the level
                        val levelArray = obj.getAsJsonArray("levels")
                        //                Creates a new hashmap to store the level and the total Experience Required to get to that level
                        val levelToExpMap =
                            HashMap<Int, Float>()
                        //                Adds the level to the map
                        for (i in 0 until levelArray.size()) {
                            val level = i + 1
                            val experience =
                                levelArray[i].getAsJsonObject()["totalExpRequired"].asFloat
                            levelToExpMap[level] = experience
                        }
                        //                Instantiates a new skyblock skill with all the data
                        val skill = SkyblockSkill(id, maxLevel, levelToExpMap)
                        //                Adds the skill to the map
                        idToSkillMap[id] = skill
                    }
                }, false, false
            )
        )
    }

    fun getSkill(skillId: String): SkyblockSkill? {
        return idToSkillMap[skillId]
    }

    //    --------------------------- Players ---------------------------
    private val playerCache = HashMap<String, SkyblockPlayer>()
    fun getPlayer(username: String): SkyblockPlayer {
        val player: SkyblockPlayer?
        if (playerCache.containsKey(username)) {
            player = playerCache[username]
            if (player?.isExpired() != true) {
                return player ?: SkyblockPlayer("") // Returns an empty player if null
            }
        } else {
            player = SkyblockPlayer(username)
        }
        player.instantiateData()
        playerCache[username] = player
        return player
    }
}
